ashpd-demo/src/portals/desktop/dynamic_launcher.rs

```diff
@@ -0,0 +1,159 @@
+            );
+        }
+
+        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
+            obj.init_template();
+        }
+    }
+    impl ObjectImpl for DynamicLauncherPage {}
+    impl WidgetImpl for DynamicLauncherPage {}
+    impl BinImpl for DynamicLauncherPage {}
+    impl PortalPageImpl for DynamicLauncherPage {}
+}
+
+glib::wrapper! {
+    pub struct DynamicLauncherPage(ObjectSubclass<imp::DynamicLauncherPage>)
+        @extends gtk::Widget, adw::Bin, PortalPage;
+}
+
+impl DynamicLauncherPage {
+    fn launcher_type(&self) -> LauncherType {
+        let imp = self.imp();
+        match imp.launcher_type_row.selected() {
+            0 => LauncherType::Application,
+            _ => LauncherType::WebApplication,
+        }
+    }
+
+    async fn select_icon(&self) -> Result<(), glib::Error> {
+        let filter = gtk::FileFilter::new();
+        filter.add_pixbuf_formats();
+
+        let filters = gio::ListStore::new::<gtk::FileFilter>();
+        filters.append(&filter);
+
+        let root = self.native().unwrap();
+        let file = gtk::FileDialog::builder()
+            .accept_label("Select")
+            .modal(true)
+            .title("Launcher Icon")
+            .filters(&filters)
+            .build()
+            .open_future(root.downcast_ref::<gtk::Window>())
+            .await?;
+
+        self.imp().icon.replace(Some(file));
+        Ok(())
+    }
+
+    async fn install(&self) -> ashpd::Result<()> {
+        let imp = self.imp();
+        if imp.icon.borrow().is_none() {
+            self.error("An icon is required");
+            return Ok(());
+        }
+
+        let root = self.native().unwrap();
+        let identifier = WindowIdentifier::from_native(&root).await;
+        let proxy = DynamicLauncherProxy::new().await?;
+
+        let launcher_name = imp.name_row.text();
+        let modal = imp.modal_switch.is_active();
+        let editable_name = imp.editable_name_switch.is_active();
+        let editable_icon = imp.editable_icon_switch.is_active();
+        let launcher_type = self.launcher_type();
+        let mut options = PrepareInstallOptions::default()
+            .modal(modal)
+            .editable_icon(editable_icon)
+            .editable_name(editable_name)
+            .launcher_type(launcher_type);
+        if launcher_type == LauncherType::WebApplication {
+            let target = imp.launcher_target_row.text();
+            options = options.target(&*target);
+        }
+
+        let selected_file = imp.icon.borrow();
+        let (data, _) = selected_file
+            .as_ref()
+            .expect("An icon is required")
+            .load_contents_future()
+            .await
+            .unwrap();
+        let icon = Icon::Bytes(data);
+
+
+        let (name, token) = proxy
+            .prepare_install(&identifier, &launcher_name, icon, options)
+            .await?
+            .response()?;
+        println!("{}", name);
+        println!("{}", token);
```

Comment:
tracing!

Reply:
it is not ready, i have a ton of local commits i didn't push yet :P